### PR TITLE
#807 meta annotations and duck typing

### PR DESCRIPTION
**Pull request by sjaakd**
As much as possible used existing functionality. Recur though nested meta-annotations, excluding already found ones. Exclude also our own and java annotations (`@Documented` and the likes).

### CODE REVIEW

**Comment by sjaakd** [comment]
> I had one question what would happen if you add a duplicated `@Mapping` (one which is defined in the meta annotation and another one on the method`?

it builds op a list of mappings, just as normal. If there are duplicates, they will be detected when adding them to the set.

**Comment by sjaakd** [comment]
> Should we add some documentation about it perhaps? I think we should say that this is experimental and that we want some feedback.

We should.. I'll add an entry..

**Comment by sjaakd** [comment]
> it builds op a list of mappings, just as normal. If there are duplicates, they will be detected when adding them to the set.

That might actually go wrong now...  I'll do some update

File: documentation/src/main/asciidoc/chapter-3-defining-a-mapper.asciidoc [hunk 1]
```diff
@@ -124,6 +124,50 @@ Collection-typed attributes with the same element type will be copied by creatin
 
 MapStruct takes all public properties of the source and target types into account. This includes properties declared on super-types.
 
+[[mapping-composition]]
+=== Mapping Composition (experimental)
+MapStruct supports the use of meta annotations. The `@Mapping` annotation can be defined with `@Target` `ElementType#ANNOTATION_TYPE`. This will allow `@Mapping` to be used on other (user defined) annotations for re-use purposes. For example:
```
Comment (chris922) [inline documentation/src/main/asciidoc/chapter-3-defining-a-mapper.asciidoc]: _can be defined_? Sounds like the user could do this.. but it actually _was defined with ..._? Following then in the next sentence: _will allow_ -> _allows_?

Comment (sjaakd) [reply documentation/src/main/asciidoc/chapter-3-defining-a-mapper.asciidoc]: true! I'll fix this.. thanks @chris922

Comment (sjaakd) [reply documentation/src/main/asciidoc/chapter-3-defining-a-mapper.asciidoc]: I rephrased it to: _The `@Mapping` annotation supports now `@Target` with `ElementType#ANNOTATION_TYPE` in addition to `ElementType#METHOD`. This allows `@Mapping` to be used ..._

**Comment by sjaakd** [comment]
> Wow, really great @sjaakd !

Thanks 😄

**Pull request merged**